r1cs/gadgets/std/src/bits/uint8.rs

```diff
@@ -338,10 +338,44 @@ impl<ConstraintF: Field> AllocGadget<u8, ConstraintF> for UInt8 {
     }
 }
 
+impl<ConstraintF: Field> CondSelectGadget<ConstraintF> for UInt8 {
+    fn conditionally_select<CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        cond: &Boolean,
+        true_value: &Self,
+        false_value: &Self,
+    ) -> Result<Self, SynthesisError> {
+        let selected_bits = true_value
+            .bits
+            .iter()
+            .zip(&false_value.bits)
+            .enumerate()
+            .map(|(i, (t, f)) | Boolean::conditionally_select(&mut cs.ns(|| format!("bit {}", i)), cond, t, f));
+        let mut bits = [Boolean::Constant(false); 8];
+        for (result, new) in bits.iter_mut().zip(selected_bits) {
+            *result = new?;
+        }
+
+        let value = cond.get_value().and_then(|cond| {
+            if cond {
+                true_value.get_value()
+            } else {
+                false_value.get_value()
+            }
+        });
+        Ok(Self { bits: bits.to_vec(), value })
+    }
+
+    fn cost() -> usize {
+       8 * <Boolean as CondSelectGadget<ConstraintF>>::cost()
+    }
+}
```

Comment:
Wouldn't it be more efficient not to select bit-wise? 
One packs both true_value and false_value as well as the UInt8 to be returned into single field elements (overall 3 C, where  C is for constraints),  and enforces a conditional select on these elements (1 C).

Reply:
It will be more costly as we would then need to enforce correct unpacking of the field gadget output of the conditionally select (since we need to output the bits). Total cost would be 4C + 8C = 12C instead of 8C